.github/workflows/main.yml

```diff
@@ -5,30 +5,59 @@ on: [push, pull_request, workflow_dispatch]
 
 jobs:
   run_test:
-    name: Test on python ${{ matrix.python-version }}
+    name: ${{ matrix.py-ver-mypy-protobuf }} / ${{matrix.py-ver-mypy-target}} / ${{matrix.py-ver-unit-tests}} protoc/mypytarget/unittests py versions
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [2.7, 3.5, 3.6, 3.7, 3.8]
+        # Running mypy-protobuf itself
+        py-ver-mypy-protobuf: [3.8.6]
+        # --python-version=X target passed into mypy
+        py-ver-mypy-target: [3.5, 2.7]
+        # Run unit tests - validating that correct typechecking usages
+        # of the google.protobuf library in various python versions
+        # also run at runtime
+        py-ver-unit-tests: [3.8.6]
+
+        include:
+          # Include py 2.7, 3.6 - 3.8 for mypy-protobuf runtime
+          - py-ver-mypy-protobuf: 3.6.12
+            py-ver-mypy-target: 3.5
+            py-ver-unit-tests: 3.8.6
+          - py-ver-mypy-protobuf: 3.7.9
+            py-ver-mypy-target: 3.5
+            py-ver-unit-tests: 3.8.6
+          - py-ver-mypy-protobuf: 2.7.18
+            py-ver-mypy-target: 3.5
+            py-ver-unit-tests: 3.8.6
+
+          # Include python runtime of google.protobuf in 2.7
+          - py-ver-mypy-protobuf: 3.8.6
+            py-ver-mypy-target: 3.5
+            py-ver-unit-tests: 2.7.18
     steps:
       - name: Install Protoc
         uses: arduino/setup-protoc@v1
         with:
-          version: '3.13.0'
+          version: '3.14.0'
       - uses: actions/checkout@v2
-      - name: Set up python ${{ matrix.python-version }}
-        uses: actions/setup-python@v2
+      - name: Install pyenv
+        uses: gabrielfalcao/pyenv-action@v7
         with:
-          python-version: ${{ matrix.python-version }}
+          # Ideally - we could specify only the versions we need, but blocked
+          # on https://github.com/gabrielfalcao/pyenv-action/issues/135
+          # versions: ${{matrix.py-ver-mypy-protobuf}}, ${{matrix.py-ver-unit-tests}}, 3.8.6
+          versions: 3.7.9, 3.6.12, 2.7.18
```

Comment:
well, you can run `pyenv install` inside run_test

Reply:
The advantage of doing it here - is that we can enable caching of the pyenv root directory - where installation is happening in an earlier stage, so have some preference to pushing installations/dependencies outside of the run_test.sh phase.

I do acknowledge that there are some dependency installations still in run_test.sh, but I'm hoping to move those out of there over time - to separate the "build" and "test" phases.